util/lua/lua.go

```diff
@@ -0,0 +1,179 @@
+
+	appv1 "github.com/argoproj/argo-cd/pkg/apis/application/v1alpha1"
+	"github.com/argoproj/argo-cd/util/settings"
+)
+
+const (
+	incorrectReturnType              = "expect table output from Lua script, not %s"
+	invalidHealthStatus              = "Lua returned an invalid health status"
+	resourceCustomizationBuiltInPath = "../../resource_customizations"
+	healthScript                     = "health.lua"
+)
+
+var (
+	box packr.Box
+)
+
+func init() {
+	box = packr.NewBox(resourceCustomizationBuiltInPath)
+}
+
+// VM Defines a struct that implements the luaVM
+type VM struct {
+	ResourceOverrides map[string]settings.ResourceOverride
+	// UseOpenLibs flag to enable open libraries. Libraries are always disabled while running, but enabled during testing to allow the use of print statements
+	UseOpenLibs bool
+}
+
+func (vm VM) runLua(obj *unstructured.Unstructured, script string) (*lua.LState, error) {
+	l := lua.NewState(lua.Options{
+		SkipOpenLibs: !vm.UseOpenLibs,
+	})
+	defer l.Close()
+	// Opens table library to allow access to functions to manulate tables
+	for _, pair := range []struct {
+		n string
+		f lua.LGFunction
+	}{
+		{lua.LoadLibName, lua.OpenPackage},
+		{lua.BaseLibName, lua.OpenBase},
+		{lua.TabLibName, lua.OpenTable},
+	} {
+		if err := l.CallByParam(lua.P{
+			Fn:      l.NewFunction(pair.f),
+			NRet:    0,
+			Protect: true,
+		}, lua.LString(pair.n)); err != nil {
+			panic(err)
+		}
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+	l.SetContext(ctx)
+	objectValue := decodeValue(l, obj.Object)
+	l.SetGlobal("obj", objectValue)
+	err := l.DoString(script)
+	return l, err
+}
+
+// ExecuteHealthLua runs the lua script to generate the health status of a resource
+func (vm VM) ExecuteHealthLua(obj *unstructured.Unstructured, script string) (*appv1.HealthStatus, error) {
+	l, err := vm.runLua(obj, script)
+	if err != nil {
+		return nil, err
+	}
+	returnValue := l.Get(-1)
+	if returnValue.Type() == lua.LTTable {
+		jsonBytes, err := luajson.Encode(returnValue)
+		if err != nil {
+			return nil, err
+		}
+		healthStatus := &appv1.HealthStatus{}
+		err = json.Unmarshal(jsonBytes, healthStatus)
+		if err != nil {
+			return nil, err
+		}
+		if !isValidHealthStatusCode(healthStatus.Status) {
+			return &appv1.HealthStatus{
+				Status:  appv1.HealthStatusUnknown,
+				Message: invalidHealthStatus,
+			}, nil
+		}
+
+		return healthStatus, nil
+	}
+	return nil, fmt.Errorf(incorrectReturnType, returnValue.Type().String())
+}
+
+// GetScript attempts to read lua script from config and then filesystem for that resource
+func (vm VM) GetHealthScript(obj *unstructured.Unstructured) (string, error) {
+	key := getConfigMapKey(obj)
+	if script, ok := vm.ResourceOverrides[key]; ok && script.HealthLua != "" {
+		return script.HealthLua, nil
+	}
+	return vm.getPredefinedLuaScripts(key, healthScript)
+}
+
+func getConfigMapKey(obj *unstructured.Unstructured) string {
```

Comment:
Related to my other comment. I think we can simply be strict on casing and drop the ToLower() comparisons. Everywhere else in our code, we compare group/kind using strict equality `==`. We can continue that assumption here.

Reply:
fixed